### PR TITLE
[llvm-lit] Add precommit test to verify current behavior of glob expansion in lit's internal shell

### PR DESCRIPTION
**Pull request by Harini0924**
This patch introduces a precommit test to verify the current behavior of glob expansion in lit's internal shell. The motivation for this test stems from an issue encountered during the BOLT test suite when running with the lit internal shell using the command:

`LIT_USE_INTERNAL_SHELL=1 ninja check-bolt`

During execution, the following error was observed:
```
File "/usr/local/google/home/harinidonthula/llvm-project/llvm/utils/lit/lit/TestRunner.py", line 416, in executeBuiltinEcho
    stdout.write(encode(maybeUnescape(args[-1])))
TypeError: string argument expected, got 'GlobItem'
```
The `executeBuiltinEcho` function in the lit testing framework expects a string to be passed to `stdout.write`, but it received a `GlobItem` object instead. This precommit test is designed to check the current behavior where the glob pattern isn't correctly expanded, leading to this `TypeError`.

While this patch doesn't fix the issue, it helps in understanding and verifying the current behavior. The feedback I received from this [PR](https://github.com/llvm/llvm-project/pull/105925) suggests using `cmd.args = expand_glob_expressions(cmd.args, shenv.cwd)` to match the behavior of `executeBuiltinMkdir` and `executeBuiltinRm`, but it is recognized that the internal shell should ideally expand globs before calling any built-in command.

**Request for Feedback:**
I'm looking for feedback on how to improve this precommit test, specifically regarding the handling and expansion of glob patterns for commands like mkdir and rm within the internal shell. Currently, the args are expanded at the beginning of these functions, which should ensure proper glob expansion. However, I'd appreciate guidance on whether I should write additional tests to verify that mkdir and rm are handling glob expansions correctly.

If such tests are recommended, I would also appreciate advice on the best approach to implement them, considering the existing framework and the way glob expansion is expected to function in the internal shell. Should these tests confirm that the current implementation passes, or are there specific edge cases I should be aware of?

**Next Steps:**

In my follow-up PR, I plan to address the UNRESOLVED error by expanding the entire command, ensuring correct and consistent behavior across all commands. The current test checks for an unresolved issue with the glob expansion, specifically looking for a `TypeError` due to an unexpanded `GlobItem`. This will be updated to reflect the correct behavior once the issue is resolved.

This change is relevant for [[RFC] Enabling the Lit Internal Shell by Default](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179/3)

### CODE REVIEW

File: llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt
Comment (arichardson) [inline llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:2]: Thanks, this is good. Can we also try testing a glob expansion passes to a non-builtin command? E.g. a local python script?

Comment (arichardson) [reply llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:2]: Could also do something like `touch %t1.txt %t2.txt` and then try `mkdir %t*.txt` and check that we get two file exists errors.

Comment (Harini0924) [reply llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:2]: Thank you for the suggestions! I have a couple of follow-up questions to ensure I'm on the right track.

For testing glob expansion with a non-built-in command, would it be appropriate to use the `find` command? I was considering using `find` to list files that match a glob pattern. Do you think this would effectively demonstrate the glob expansion, or is there another command you would recommend?

Regarding the mkdir suggestion, I'm a bit confused about where this test should be placed. Should it go in the same file as the glob-echo test (`llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt`), or would it be better to create a new file specifically for mkdir, such as `llvm/utils/lit/tests/Inputs/shtest-glob/glob-mkdir.txt`? And then, in `llvm/utils/lit/tests/shtest-glob.py`, should this test be expected to fail, indicating the current behavior before implementing the fix?

File: llvm/utils/lit/tests/Inputs/shtest-glob/example_file1.txt
Comment (arichardson) [inline llvm/utils/lit/tests/Inputs/shtest-glob/example_file1.txt:2]:
```suggestion
```
This file should not contain any RUN: lines especially not ones that cannot actually be run. If you rename it to something like `example_file1.input` lit will not try to run it (since we have `config.suffixes = [".txt"]`).

File: llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt
Comment (arichardson) [inline llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt:2]:
```suggestion
```
same here - also rename it to have a different suffix.

File: llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt
Comment (arichardson) [inline llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:2]:
```suggestion
# RUN: echo %S/example_file*.input
```

Comment (arichardson) [reply llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt:2]: I'm not too familiar with the structure of the lit internal test suite - I'd say keeping the tests in separate files matches the current layout more closely. Looking at the code mkdir handles globs already so if you just run something like `not mkdir %S/example_input%.input` you can verify the error output.

**Review by arichardson — APPROVED** [review]

**Review by arichardson — CHANGES_REQUESTED** [review]
(accidentally accepted)

**Comment by Harini0924** [comment]
I will address the code formatter issues later.

File: llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt
Comment (Harini0924) [reply llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt]: I wasn’t too sure when you suggested renaming the suffix. I assumed you meant renaming this file suffix to `.input` to match the other file, so I did that. However, if you intended for a different suffix, like `.data` or something else, I can make that change. Please let me know if you have a specific suffix in mind.

File: llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt
Comment (Harini0924) [reply llvm/utils/lit/tests/Inputs/shtest-glob/glob-echo.txt]: done.

**Comment by Harini0924** [comment]
@arichardson I followed your suggestion about keeping the tests in separate files to align with the current structure of the lit internal test suite. I created a file `glob-mkdir.txt` that checks if the mkdir command fails as expected when attempting to create a directory that already exists (using `example_file`). This is reflected in the `shtest-glob.py file`, where the test is expected to fail. Is this what you had in mind? Should I add or change anything else?

**Review by ilovepi — COMMENTED** [review]
LGTM, as long as @arichardson is satisfied.

File: llvm/utils/lit/tests/shtest-glob.py
Comment (arichardson) [inline llvm/utils/lit/tests/shtest-glob.py:12]: I was surprised there is no more specific error here but it looks like the builtin mkdir does not print any output right now. Turns out it doesn't handle EEXISTS the same way as the real mkdir command. I can submit a patch for this once this commit lands.

**Review by arichardson — APPROVED** [review]
Thanks, this looks good to me.

**Comment by arichardson** [comment]
> Thanks, this looks good to me.

But please do fix the trailing whitespace before committing.

File: llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt
Comment (arichardson) [reply llvm/utils/lit/tests/Inputs/shtest-glob/example_file2.txt]: Yes that's exactly what I meant. The suffix doesn't matter we just want two files matching the glob.

File: llvm/utils/lit/tests/shtest-glob.py
Comment (Harini0924) [reply llvm/utils/lit/tests/shtest-glob.py:12]: A patch to address that would be great, and I’m looking forward to seeing it once this commit lands.

**Review by arichardson — APPROVED** [review]

**Pull request merged**
